File: api/internal/api/estado_del_webhook.go
package api

import (
	"net/http"
	"strconv"
	"strings"
	"time"

	"procovar/reparto-api/internal/httpx"
	"procovar/reparto-api/internal/store/sqlc"
)

// CÓMO VA EL WEBHOOK, EN LOS DOS SENTIDOS. Sólo para administración.
//
// Jose, 26/09/2026: «aparte de recibir y enviar, tener también para ver cómo está
// funcionando el webhook, tanto lo que envía como lo que recibe» y «de salida también, para
// saber si llegó correcto o no». Y quién puede mirarlo: «esto es para administración, esta
// vista no la puede ver nadie».
//
// POR QUÉ ES DE ADMINISTRACIÓN Y NO DE TODOS. Esto no son datos de una sucursal: es el
// estado de una tubería entre dos sistemas, con motivos de error de PEDIDO dentro. A quien
// arma rutas no le sirve de nada y sólo puede confundirle —«hay 12 rechazados» no es una
// tarea suya—. Va detrás de `ExigirAdmin`, que es el mismo listón del resto de lo
// administrativo, y **no se pone en el menú**.
//
// LO QUE CONTESTA, y son tres preguntas distintas que la gente confunde:
//
//  1. **¿Respira?** Las horas de la última tanda de cada lado. Un webhook que lleva seis
//     horas sin recibir nada no da ningún error: sólo deja de pasar cosas. `null` es
//     «nunca», que no es «hace mucho» y se dice con otras palabras.
//  2. **¿Llega lo que sale?** Cada tanda enviada con su código HTTP y su motivo literal.
//     Un aviso sin enviar puede ser «PEDIDO está caído» —la llamada no llegó— o «PEDIDO lo
//     rechazó» —llegó perfectamente y dijo que no—: desde fuera se ven igual.
//  3. **¿Se escribe lo que entra?** Cada tanda recibida con cuántos traía y cuántos se
//     escribieron. Cuando un pedido no aparece en el reparto, esto dice de quién es el
//     problema.

// TopeDelEstadoDelWebhook: cuántas líneas de cada lado se devuelven.
//
// Esto se mira para responder «¿qué pasó hace un rato?», no para auditar el mes. Con
// veinte de cada lado se ve el ritmo; con mil, la pantalla deja de leerse y la consulta
// empieza a pesar.
const TopeDelEstadoDelWebhook = 20

// TopeMaximoDelEstadoDelWebhook: hasta dónde se deja subir con `?tope=`.
const TopeMaximoDelEstadoDelWebhook = 200

type EstadoDelWebhookSalida struct {
	Resumen   ResumenDelWebhookSalida `json:"resumen"`
	Enviados  []EnvioDelWebhookSalida `json:"enviados"`
	Recibidos []RecepcionDelWebhook   `json:"recibidos"`
	SinMandar []AvisoSinMandar        `json:"sinMandar"`

	// LA CUARTA PREGUNTA, desde el 26/09/2026: **¿se está midiendo desde donde hay que
	// medir?** Va aquí y no en una pantalla nueva porque es la misma tubería —lo que PEDIDO
	// manda y lo que este lado hace con ello— y porque es de administración por el mismo
	// motivo: no es un dato de una sucursal, es el estado de un acuerdo entre dos sistemas.
	//
	// Lo que pasa cuando esto no está: la distancia del domicilio se mide DESDE EL ALMACÉN, y
	// hasta hoy había uno por sucursal. Ahora hay varios, y en Santiago **dos de cada tres
	// pedidos salen de AURORA** mientras el principal es PV-STGO. Un pedido medido desde el
	// principal porque su almacén no estaba dado de alta tiene el mismo aspecto que uno bien
	// medido: mismos decimales, misma tarjeta. Sin esta lista, la única forma de enterarse es
	// que alguien de PEDIDO cuente los renglones por almacén, que es como se descubrió.
	SinMedir []AlmacenSinMedirSalida `json:"sinMedir"`
	// SinMedirTruncado: si el tope se comió filas. Ver `CLAUDE.md` §3 — pedir un tope y no
	// mirar si se alcanzó es el fallo que más caro sale aquí, y esta lista está ORDENADA POR
	// CUENTA DESCENDENTE, así que lo que el tope se come son los almacenes de uno o dos
	// pedidos: precisamente el que acaba de aparecer y hay que cazar temprano.
	SinMedirTruncado bool `json:"sinMedirTruncado"`

	ServidorAt time.Time `json:"servidorAt"`
}

// AlmacenSinMedirSalida es una fila de «no se midió desde su almacén»: qué llegó, por qué no
// se pudo usar, cuántos pedidos y desde cuándo.
//
// CADA MOTIVO SE ARREGLA EN UN SITIO DISTINTO y por eso viaja: dar de alta el almacén en
// Accesos, ponerle el punto, ponerle el código, o que PEDIDO empiece a mandar el campo. Con un
// «hay 300 pedidos mal medidos» sin motivo, nadie sabe qué hacer.
type AlmacenSinMedirSalida struct {
	SucursalCodigo *string    `json:"sucursalCodigo"`
	Motivo         *string    `json:"motivo"`
	Codigo         *string    `json:"codigo"`
	Nombre         *string    `json:"nombre"`
	Pedidos        int64      `json:"pedidos"`
	Desde          *time.Time `json:"desde"`
	Hasta          *time.Time `json:"hasta"`
}

type ResumenDelWebhookSalida struct {
	// `null` = NUNCA. No es lo mismo que «hace mucho» y la pantalla lo dice distinto.
	UltimaEntrada *time.Time `json:"ultimaEntrada"`
	UltimaSalida  *time.Time `json:"ultimaSalida"`

	EscritosHoy        int64 `json:"escritosHoy"`
	RechazadosAlEntrar int64 `json:"rechazadosAlEntrar"`

	AvisosPendientes int64 `json:"avisosPendientes"`
	AvisosRechazados int64 `json:"avisosRechazados"`
	// El más viejo sin mandar. Es EL número que dice si esto está atascado: veinte
	// pendientes de hace un minuto es el ritmo normal; uno de hace seis horas, no.
	PendienteMasViejo *time.Time `json:"pendienteMasViejo"`
}

type EnvioDelWebhookSalida struct {
	Destino    string    `json:"destino"`
	Mandados   int32     `json:"mandados"`
	Aceptados  int32     `json:"aceptados"`
	Rechazados int32     `json:"rechazados"`
	HTTP       *int32    `json:"http"`
	Motivo     *string   `json:"motivo"`
	DuracionMs int32     `json:"duracionMs"`
	CreatedAt  time.Time `json:"createdAt"`
}

type RecepcionDelWebhook struct {
	Origen     string    `json:"origen"`
	Traidos    int32     `json:"traidos"`
	Escritos   int32     `json:"escritos"`
	Rechazados int32     `json:"rechazados"`
	Motivos    *string   `json:"motivos"`
	DuracionMs int32     `json:"duracionMs"`
	CreatedAt  time.Time `json:"createdAt"`
}

// AvisoSinMandar: los que no llegaron, con su motivo. Es la bandeja que alguien tiene que
// mirar — un rechazado no se reintenta solo y se queda ahí a propósito.
type AvisoSinMandar struct {
	PedidoID  string     `json:"pedidoId"`
	Folio     *string    `json:"folio"`
	Estado    string     `json:"estado"`
	Situacion string     `json:"situacion"`
	Intentos  int32      `json:"intentos"`
	Motivo    *string    `json:"motivo"`
	OcurrioAt *time.Time `json:"ocurrioAt"`
	CreatedAt time.Time  `json:"createdAt"`
}

// GET /api/admin/webhook
func (s *Servidor) estadoDelWebhook(w http.ResponseWriter, r *http.Request) {
	a, ok := acotado(w, r)
	if !ok {
		return
	}
	tope := int32(TopeDelEstadoDelWebhook)
	if n, err := strconv.Atoi(strings.TrimSpace(r.URL.Query().Get("tope"))); err == nil &&
		n > 0 && n <= TopeMaximoDelEstadoDelWebhook {
		tope = int32(n)
	}

	res, err := a.ResumenDelWebhook(r.Context())
	if err != nil {
		httpx.ErrorInterno(w, r, err)
		return
	}
	envios, err := a.ListarEnviosDelWebhook(r.Context(), tope)
	if err != nil {
		httpx.ErrorInterno(w, r, err)
		return
	}
	recepciones, err := a.ListarRecepcionesDelWebhook(r.Context(), tope)
	if err != nil {
		httpx.ErrorInterno(w, r, err)
		return
	}
	// LOS QUE NO LLEGARON, los dos motivos juntos: pendientes y rechazados. Es la bandeja
	// que alguien tiene que mirar, y se pide sin filtro para que salgan los dos —filtrar
	// por uno dejaría la mitad del problema fuera de la pantalla.
	sinMandar, err := a.ListarAvisosAPedido(r.Context(), sqlc.ListarAvisosAPedidoParams{
		Tope: tope,
	})
	if err != nil {
		httpx.ErrorInterno(w, r, err)
		return
	}

	// LO QUE NO SE MIDIÓ DESDE SU ALMACÉN, acotado a lo que esta persona puede ver.
	//
	// SE PIDE UNA FILA DE MÁS. Es lo que convierte «truncado» en un dato y no en una sospecha:
	// si vuelve la de más, hay más; si no vuelve, ésta era toda la lista. Es el molde de
	// `tablero.go`, y aquí hace más falta que allí porque la lista va ordenada por cuenta
	// descendente: sin esto, el almacén que aparece hoy con un pedido —el que hay que cazar
	// antes de que sean mil— se cae por debajo del corte y nadie lo echa en falta.
	//
	// EL ALCANCE SALE DE QUIÉN PREGUNTA y no de la cabecera: `nil` sólo para quien ve las
	// ocho. Con `CodigosDeSucursalesVisibles` un logístico de Camagüey ve lo suyo y nada más
	// — la regla 1 de la casa, la que en delivery dejó a un operador de Santiago viendo los
	// precios de La Habana.
	var visibles []string
	if !a.Todas() {
		codigos, err := a.CodigosDeSucursalesVisibles(r.Context())
		if err != nil {
			httpx.ErrorInterno(w, r, err)
			return
		}
		// Se arranca en `[]string{}` y NO en nil: un nil significa «todas» en la consulta, así
		// que una persona cuya sucursal no tenga código vería las ocho. Con la lista vacía no
		// ve ninguna, que es la verdad y el lado seguro.
		visibles = []string{}
		for _, c := range codigos {
			if c.ExternalID != nil && strings.TrimSpace(*c.ExternalID) != "" {
				visibles = append(visibles, strings.TrimSpace(*c.ExternalID))
			}
		}
	}
	sinMedir, err := a.AlmacenesDelPedidoSinMedir(r.Context(), visibles, tope+1)
	if err != nil {
		httpx.ErrorInterno(w, r, err)
		return
	}
	huboMas := int32(len(sinMedir)) > tope
	if huboMas {
		sinMedir = sinMedir[:tope]
	}

	salida := EstadoDelWebhookSalida{
		Resumen: ResumenDelWebhookSalida{
			UltimaEntrada:      hora(res.UltimaEntrada),
			UltimaSalida:       hora(res.UltimaSalida),
			EscritosHoy:        res.EscritosHoy,
			RechazadosAlEntrar: res.RechazadosAlEntrar,
			AvisosPendientes:   res.AvisosPendientes,
			AvisosRechazados:   res.AvisosRechazados,
			PendienteMasViejo:  hora(res.PendienteMasViejo),
		},
		Enviados:         make([]EnvioDelWebhookSalida, 0, len(envios)),
		Recibidos:        make([]RecepcionDelWebhook, 0, len(recepciones)),
		SinMandar:        make([]AvisoSinMandar, 0, len(sinMandar)),
		SinMedir:         make([]AlmacenSinMedirSalida, 0, len(sinMedir)),
		SinMedirTruncado: huboMas,
		ServidorAt:       time.Now().UTC(),
	}
	for _, x := range sinMedir {
		salida.SinMedir = append(salida.SinMedir, AlmacenSinMedirSalida{
			SucursalCodigo: x.SucursalCodigo, Motivo: x.Motivo,
			Codigo: x.Codigo, Nombre: x.Nombre, Pedidos: x.Cuantos,
			Desde: hora(x.Desde), Hasta: hora(x.Hasta),
		})
	}
	for _, e := range envios {
		salida.Enviados = append(salida.Enviados, EnvioDelWebhookSalida{
			Destino: e.Destino, Mandados: e.Mandados, Aceptados: e.Aceptados,
			Rechazados: e.Rechazados, HTTP: e.Http, Motivo: e.Motivo,
			DuracionMs: e.DuracionMs, CreatedAt: e.CreatedAt.Time,
		})
	}
	for _, c := range recepciones {
		salida.Recibidos = append(salida.Recibidos, RecepcionDelWebhook{
			Origen: c.Origen, Traidos: c.Traidos, Escritos: c.Escritos,
			Rechazados: c.Rechazados, Motivos: c.Motivos,
			DuracionMs: c.DuracionMs, CreatedAt: c.CreatedAt.Time,
		})
	}
	for _, v := range sinMandar {
		if v.Situacion == sqlc.AvisoAPedidoEstadoEnviado {
			continue
		}
		salida.SinMandar = append(salida.SinMandar, AvisoSinMandar{
			PedidoID: v.PedidoID, Folio: v.Folio, Estado: v.Estado,
			Situacion: string(v.Situacion), Intentos: v.Intentos, Motivo: v.Motivo,
			OcurrioAt: hora(v.OcurrioAt), CreatedAt: v.CreatedAt.Time,
		})
	}
	httpx.JSON(w, r, http.StatusOK, salida)
}

File: app/lib/pantallas/webhook/datos/estado_del_webhook.dart
import 'package:reparto/nucleo/red/cliente_api.dart';

/// CÓMO VA EL CANAL CON PEDIDO, leído del servidor.
///
/// Esta pantalla **sólo existe en la web y sólo la ve el desarrollador**. Jose, 26/09/2026:
/// «eso me lo dejas en la web solamente, no lo pongas en más ningún lado» y «que sólo lo
/// pueda ver yo, eso no lo puede ver más nadie, sólo yo, el desarrollador».
///
/// LAS PALABRAS SON LAS MISMAS QUE LAS DE PEDIDO, y eso no es un detalle: la otra mitad de
/// este tubo tiene su propia pantalla en PEDIDO → Sincronización → Reparto. Si una dice
/// «pendientes» y la otra «esperando», comparar las dos obliga a traducir, y en una
/// traducción es donde se cuela la conclusión equivocada. Se acordaron entre las dos
/// sesiones el 26/09/2026:
///
///   **esperando** · **sin terminar** · **el más viejo** · **último aviso**
///
/// Con dos diferencias a propósito: aquí no hay «nadie está leyendo» —el que lee es éste— y
/// aquí sí está «el más viejo sin mandar», que es EL número que dice si la salida está
/// atascada: veinte esperando de hace un minuto es el ritmo normal, uno de hace seis horas
/// no lo es.
class EstadoDelWebhook {
  const EstadoDelWebhook({
    required this.resumen,
    required this.enviados,
    required this.recibidos,
    required this.sinMandar,
  });

  factory EstadoDelWebhook.deJson(Map<String, Object?> j) => EstadoDelWebhook(
    resumen: ResumenDelWebhook.deJson(
      (j['resumen'] as Map<String, Object?>?) ?? const {},
    ),
    enviados: [
      for (final e in (j['enviados'] as List<Object?>? ?? const []))
        if (e is Map<String, Object?>) TandaEnviada.deJson(e),
    ],
    recibidos: [
      for (final e in (j['recibidos'] as List<Object?>? ?? const []))
        if (e is Map<String, Object?>) TandaRecibida.deJson(e),
    ],
    sinMandar: [
      for (final e in (j['sinMandar'] as List<Object?>? ?? const []))
        if (e is Map<String, Object?>) AvisoSinMandar.deJson(e),
    ],
  );

  final ResumenDelWebhook resumen;
  final List<TandaEnviada> enviados;
  final List<TandaRecibida> recibidos;
  final List<AvisoSinMandar> sinMandar;
}

class ResumenDelWebhook {
  const ResumenDelWebhook({
    this.ultimaEntrada,
    this.ultimaSalida,
    this.escritosHoy = 0,
    this.rechazadosAlEntrar = 0,
    this.esperando = 0,
    this.rechazados = 0,
    this.masViejo,
  });

  factory ResumenDelWebhook.deJson(Map<String, Object?> j) => ResumenDelWebhook(
    ultimaEntrada: _fecha(j['ultimaEntrada']),
    ultimaSalida: _fecha(j['ultimaSalida']),
    escritosHoy: _entero(j['escritosHoy']),
    rechazadosAlEntrar: _entero(j['rechazadosAlEntrar']),
    esperando: _entero(j['avisosPendientes']),
    rechazados: _entero(j['avisosRechazados']),
    masViejo: _fecha(j['pendienteMasViejo']),
  );

  /// `null` = **nunca**, que no es lo mismo que «hace mucho» y la pantalla lo dice con
  /// otras palabras. Un canal que nunca ha recibido nada y uno que lleva seis horas
  /// parado se arreglan en sitios distintos.
  final DateTime? ultimaEntrada;
  final DateTime? ultimaSalida;

  final int escritosHoy;
  final int rechazadosAlEntrar;

  /// Avisos nuestros que aún no han salido hacia PEDIDO.
  final int esperando;

  /// Los que PEDIDO rechazó: llegaron perfectamente y dijo que no. **No se reintentan**, y
  /// por eso son una bandeja que alguien tiene que mirar, no un número que baja solo.
  final int rechazados;

  /// El más viejo sin mandar. Es EL número que dice si esto está atascado.
  final DateTime? masViejo;

  /// ¿Hay algo que mirar? Se usa para el punto de color, y NO incluye `esperando`: veinte
  /// esperando de hace un minuto es el ritmo normal de un cierre de ruta.
  bool get hayQueMirar => rechazados > 0 || estaAtascado;

  /// Atascado: algo lleva más de diez minutos sin salir. Es el mismo listón que usa la
  /// pantalla de PEDIDO, acordado entre las dos sesiones para que las dos digan lo mismo.
  bool get estaAtascado {
    final desde = masViejo;
    if (desde == null) return false;
    return DateTime.now().difference(desde) > const Duration(minutes: 10);
  }
}

/// Una tanda que salió hacia PEDIDO.
class TandaEnviada {
  const TandaEnviada({
    required this.mandados,
    required this.aceptados,
    required this.rechazados,
    required this.cuando,
    this.http,
    this.motivo,
    this.duracionMs = 0,
  });

  factory TandaEnviada.deJson(Map<String, Object?> j) => TandaEnviada(
    mandados: _entero(j['mandados']),
    aceptados: _entero(j['aceptados']),
    rechazados: _entero(j['rechazados']),
    http: j['http'] == null ? null : _entero(j['http']),
    motivo: j['motivo'] as String?,
    duracionMs: _entero(j['duracionMs']),
    cuando: _fecha(j['createdAt']) ?? DateTime.now(),
  );

  final int mandados;
  final int aceptados;
  final int rechazados;

  /// El código tal cual. Un 200 con cero aceptados y un 502 **no son lo mismo**: uno llegó
  /// y dijo que no, el otro no llegó. Guardar sólo «falló» los confunde.
  final int? http;
  final String? motivo;
  final int duracionMs;
  final DateTime cuando;

  bool get llegoBien => http != null && http! >= 200 && http! < 300;
}

/// Una tanda que entró, por la cola o por HTTP.
class TandaRecibida {
  const TandaRecibida({
    required this.origen,
    required this.traidos,
    required this.escritos,
    required this.rechazados,
    required this.cuando,
    this.motivos,
    this.duracionMs = 0,
  });

  factory TandaRecibida.deJson(Map<String, Object?> j) => TandaRecibida(
    origen: (j['origen'] as String?) ?? 'pedido',
    traidos: _entero(j['traidos']),
    escritos: _entero(j['escritos']),
    rechazados: _entero(j['rechazados']),
    motivos: j['motivos'] as String?,
    duracionMs: _entero(j['duracionMs']),
    cuando: _fecha(j['createdAt']) ?? DateTime.now(),
  );

  /// De qué puerta vino. **SON TRES, y se dicen con palabras distintas**:
  ///
  ///  - `webhook` — un aviso de PEDIDO por `POST /api/webhooks/pedido`, de uno en uno.
  ///  - `stream` — por la cola de Redis, mientras las dos puertas convivan.
  ///  - `pedido` — una tanda del espejo por `/api/quote/batch`, de 200 en 200.
  ///
  /// LOS TRES EN LA MISMA LISTA: son caminos para lo mismo, y tenerlos separados obliga a
  /// mirar en tres sitios para responder «¿está entrando algo?», que es la primera pregunta
  /// cuando algo no llega.
  ///
  /// PERO CON NOMBRES DISTINTOS, y eso costó una vuelta: `webhook` y `pedido` se llamaban los
  /// dos `pedido`, así que los avisos —de uno en uno— quedaban enterrados entre lotes de
  /// doscientos y no se podía contestar «¿entra algo POR EL WEBHOOK?». Se vio en producción el
  /// 26/09/2026 con las dos clases de fila mezcladas en la tabla.
  final String origen;

  final int traidos;

  /// Los que llevaron a una acción.
  final int escritos;

  /// Los que NO llevaron a nada. **No son fallos**: un pedido repetido en la misma tanda,
  /// o un traer que un borrado anuló, es el sistema haciendo lo correcto. Pintarlos de
  /// rojo enseña a ignorar el rojo, y entonces el rojo de verdad tampoco se mira.
  final int rechazados;

  final String? motivos;
  final int duracionMs;
  final DateTime cuando;

  String get comoSeLlama => switch (origen) {
    'webhook' => 'por el webhook',
    'stream' => 'por la cola',
    'pedido' => 'por el lote del espejo',
    // UN ORIGEN QUE NO SE CONOCE SE DICE TAL CUAL, no se traduce a «por HTTP».
    // Traducirlo escondería que hay una puerta nueva que nadie sabe que existe.
    _ => origen,
  };
}

/// Un aviso nuestro que no llegó a PEDIDO.
class AvisoSinMandar {
  const AvisoSinMandar({
    required this.pedidoId,
    required this.estado,
    required this.situacion,
    required this.intentos,
    required this.cuando,
    this.folio,
    this.motivo,
  });

  factory AvisoSinMandar.deJson(Map<String, Object?> j) => AvisoSinMandar(
    pedidoId: (j['pedidoId'] as String?) ?? '',
    folio: j['folio'] as String?,
    estado: (j['estado'] as String?) ?? '',
    situacion: (j['situacion'] as String?) ?? '',
    intentos: _entero(j['intentos']),
    motivo: j['motivo'] as String?,
    cuando: _fecha(j['createdAt']) ?? DateTime.now(),
  );

  final String pedidoId;
  final String? folio;
  final String estado;

  /// `pendiente` = no se pudo ni preguntar, se reintenta. `rechazado` = PEDIDO dijo que
  /// no, **y eso no se reintenta**: repetir lo mismo da lo mismo. Son dos cosas distintas
  /// y se arreglan distinto.
  final String situacion;

  final int intentos;
  final String? motivo;
  final DateTime cuando;

  bool get loRechazaron => situacion == 'rechazado';
}

DateTime? _fecha(Object? v) =>
    v is String ? DateTime.tryParse(v)?.toLocal() : null;

int _entero(Object? v) => switch (v) {
  final int n => n,
  final num n => n.toInt(),
  final String s => int.tryParse(s) ?? 0,
  _ => 0,
};

/// Lee el estado del canal. **Contra el servidor y sólo contra el servidor**: esto no vive
/// en la copia local, que es de la sucursal y esto no es de ninguna.
class RepositorioDelWebhook {
  const RepositorioDelWebhook(this._api);

  final ClienteApi _api;

  Future<EstadoDelWebhook> mirar() async {
    final crudo = await _api.pedir<Map<String, Object?>>('/admin/webhook');
    return EstadoDelWebhook.deJson(crudo);
  }
}
